alf/drivers/async_off_policy_driver.py

```diff
@@ -108,8 +110,15 @@ def __init__(self,
                 tf.summary.experimental.get_step(). If this is still None, a
                 counter will be created.
         """
+        if isinstance(env_or_env_fn, TFEnvironment):
```

Comment:
Can we just enforce env_or_env_fn to be env? If this is more efficient, then probably there will be no need for an env_fn.

Reply:
total env num depends on `num_parallel_environments`  and  `AsyncOffPolicyDriver.num_envs`  

so  we should move  configurable parameter `num_envs` to `TrainerConfig`, is it appropriate ?

Reply:
I mean we can directly discard the second branch here and just assume env_or_env_fn is always an environment for efficiency. num_envs stays unchanged.

Reply:
env can be created by other functions, `create_environment` is just the default choice 

eg: 
https://github.com/HorizonRobotics/alf/blob/62b3aa1158b7162fac66c7ad568ba59a88266e36/alf/drivers/off_policy_driver_test.py#L112-L118
pass an env_fn

Reply:
OK, never mind.